File: app/src/main/java/com/qyt/accumulation/util/SPUtil.java
package com.qyt.accumulation.util;

import android.content.SharedPreferences;
import android.text.TextUtils;

import com.qyt.accumulation.App;
import com.qyt.accumulation.entity.User;

/**
 * SharedPreferences 工具类
 * Created by Relish on 2016/12/3.
 */

public class SPUtil {
    private static final String TAG = "SPUtil";

    private static SharedPreferences settings;

    static {
        settings = App.CONTEXT.getSharedPreferences("android_preference", 0);
    }

    public static User getUser() {
        User user = new User();
        user.setName(checkNull("name", user.getName()));
        user.setPassword(checkNull("password", user.getPassword()));
        user.setPhoto(checkNull("photo", user.getPhoto()));
        user.setMobile(checkNull("mobile", user.getMobile()));
        return user;
    }


    public static void saveUser(User user) {
        if (user == null) {
            AppLog.e(TAG, "saveUser", "coming param is null!!!");
            return;
        }
        save("name", user.getName());
        save("photo", user.getPhoto());
        save("password", user.getPassword());
        save("mobile", user.getMobile());
    }


    private static void save(String key, Object value) {
        if (value instanceof String) {
            //noinspection ConstantConditions
            putString(key, checkNull(key, (String) value));
        } else if (value instanceof Integer) {
            //noinspection ConstantConditions
            putInt(key, checkNull(key, (Integer) value));
        } else if (value instanceof Long) {
            //noinspection ConstantConditions
            putLong(key, checkNull(key, (Long) value));
        } else if (value instanceof Float) {
            //noinspection ConstantConditions
            putFloat(key, checkNull(key, (Float) value));
        } else if (value instanceof Boolean) {
            //noinspection ConstantConditions
            putBoolean(key, checkNull(key, (Boolean) value));
        } else {
            if (value != null) {
                AppLog.e(TAG, "save", "Unknown type: " + value.getClass().toString());
            }
        }
    }

    private static <T> T checkNull(String key, T value) {
        if (value instanceof String) {
            //noinspection unchecked
            return TextUtils.isEmpty((String) value) ||
                    ((String) value).equalsIgnoreCase("null") ?
                    (T) getString(key, "") : value;
        } else if (value instanceof Integer || value instanceof Boolean ||
                value instanceof Long || value instanceof Float) {
            //noinspection unchecked
            return value;
        } else {
            if (value != null)
                AppLog.e(TAG, "checkNull", "Unknown type: " + value.getClass());
            return null;
        }
    }

    public static boolean putString(String key, String value) {
        if (key != null && value != null) {
            SharedPreferences.Editor editor = settings.edit();
            editor.putString(key, value);
            return editor.commit();
        } else {
            return false;
        }
    }

    public static String getString(String key) {
        return getString(key, (String) null);
    }

    public static String getString(String key, String defaultValue) {
        return settings.getString(key, defaultValue);
    }

    public static boolean putInt(String key, int value) {
        SharedPreferences.Editor editor = settings.edit();
        editor.putInt(key, value);
        return editor.commit();
    }

    public static int getInt(String key) {
        return getInt(key, -1);
    }

    public static int getInt(String key, int defaultValue) {
        return settings.getInt(key, defaultValue);
    }

    public static boolean putLong(String key, long value) {
        SharedPreferences.Editor editor = settings.edit();
        editor.putLong(key, value);
        return editor.commit();
    }

    public static long getLong(String key) {
        return getLong(key, -1L);
    }

    public static long getLong(String key, long defaultValue) {
        return settings.getLong(key, defaultValue);
    }

    public static boolean putFloat(String key, float value) {
        SharedPreferences.Editor editor = settings.edit();
        editor.putFloat(key, value);
        return editor.commit();
    }

    public static float getFloat(String key) {
        return getFloat(key, -1.0F);
    }

    public static float getFloat(String key, float defaultValue) {
        return settings.getFloat(key, defaultValue);
    }

    public static boolean putBoolean(String key, boolean value) {
        SharedPreferences.Editor editor = settings.edit();
        editor.putBoolean(key, value);
        return editor.commit();
    }

    public static boolean getBoolean(String key) {
        return getBoolean(key, false);
    }

    public static boolean getBoolean(String key, boolean defaultValue) {
        return settings.getBoolean(key, defaultValue);
    }

    public static void clearAll() {
        SharedPreferences.Editor editor = settings.edit();
        editor.clear().apply();
    }


}

File: app/src/main/java/wang/relish/accumulation/util/SPUtil.kt
package wang.relish.accumulation.util

import android.content.Context
import android.content.SharedPreferences
import android.text.TextUtils

import wang.relish.accumulation.App
import wang.relish.accumulation.entity.User

/**
 * SharedPreferences 工具类
 * Created by Relish on 2016/12/3.
 */

object SPUtil {
    private val TAG = "SPUtil"

    private var settings: SharedPreferences? = null

    init {
        settings = App.CONTEXT!!.getSharedPreferences("user", Context.MODE_PRIVATE)
    }

    val user: User
        get() {
            val user = User()
            user.name = getString("name", "")
            user.password = getString("password", "")
            user.photo = getString("photo", "")
            user.mobile = getString("mobile", "")
            return user
        }


    fun saveUser(user: User?) {
        if (user == null) {
            AppLog.e(TAG, "saveUser", "coming param is null!!!")
            return
        }
        save("name", user.name)
        save("photo", user.photo)
        save("password", user.password)
        save("mobile", user.mobile)
    }


    private fun save(key: String, value: Any?) {
        if (value is String) {

            putString(key, checkNull(key, value as String?))
        } else if (value is Int) {

            putInt(key, checkNull(key, value as Int?)!!)
        } else if (value is Long) {

            putLong(key, checkNull(key, value as Long?)!!)
        } else if (value is Float) {

            putFloat(key, checkNull(key, value as Float?)!!)
        } else if (value is Boolean) {

            putBoolean(key, checkNull(key, value as Boolean?)!!)
        } else {
            if (value != null) {
                AppLog.e(TAG, "save", "Unknown type: " + value.javaClass.toString())
            }
        }
    }

    private fun <T : Any> checkNull(key: String, value: T?): T? {
        if (value is String) {

            return if (TextUtils.isEmpty(value as String?) || value.equals("null", ignoreCase = true))
                getString(key, "") as T
            else
                value
        } else if (value is Int || value is Boolean ||
                value is Long || value is Float) {

            return value
        } else {
            if (value != null)
                AppLog.e(TAG, "checkNull", "Unknown type: " + value.javaClass)
            return null
        }
    }

    fun putString(key: String?, value: String?): Boolean {
        if (key != null && value != null) {
            val editor = settings!!.edit()
            editor.putString(key, value)
            return editor.commit()
        } else {
            return false
        }
    }

    @JvmOverloads fun getString(key: String, defaultValue: String = (null as String?)!!): String {
        return settings!!.getString(key, defaultValue)
    }

    fun putInt(key: String, value: Int): Boolean {
        val editor = settings!!.edit()
        editor.putInt(key, value)
        return editor.commit()
    }

    @JvmOverloads fun getInt(key: String, defaultValue: Int = -1): Int {
        return settings!!.getInt(key, defaultValue)
    }

    fun putLong(key: String, value: Long): Boolean {
        val editor = settings!!.edit()
        editor.putLong(key, value)
        return editor.commit()
    }

    @JvmOverloads fun getLong(key: String, defaultValue: Long = -1L): Long {
        return settings!!.getLong(key, defaultValue)
    }

    fun putFloat(key: String, value: Float): Boolean {
        val editor = settings!!.edit()
        editor.putFloat(key, value)
        return editor.commit()
    }

    @JvmOverloads fun getFloat(key: String, defaultValue: Float = -1.0f): Float {
        return settings!!.getFloat(key, defaultValue)
    }

    fun putBoolean(key: String, value: Boolean): Boolean {
        val editor = settings!!.edit()
        editor.putBoolean(key, value)
        return editor.commit()
    }

    @JvmOverloads fun getBoolean(key: String, defaultValue: Boolean = false): Boolean {
        return settings!!.getBoolean(key, defaultValue)
    }

    fun clearAll() {
        val editor = settings!!.edit()
        editor.clear().apply()
    }


}
